common/Unit.hpp

```diff
@@ -433,4 +434,70 @@ class UnitKit : public UnitBase
     }
 };
 
+/// Enum macro specifically for state-machines.
+/// Has several limitations, some intentional. For example,
+/// the states must have automatic, sequential, values.
+/// But also has some advantages, for example it can be used inside classes.
+/// Some ideas from https://stackoverflow.com/questions/28828957/enum-to-string-in-modern-c11-c14-c17-and-future-c20
+/// and from https://github.com/pfultz2/Cloak/wiki/C-Preprocessor-tricks,-tips,-and-idioms
+
+#define STRINGIFY(NAME, e) #NAME "::" #e,
+#define CONCAT(X, Y) X##Y
+#define CALL(X, ...) X(__VA_ARGS__)
+
+#define APPLY1(MACRO, NAME, e) MACRO(NAME, e)
```

Comment:
Are you sure this is impossible to do with variadic templates? Introducing new macros in C++ code is always a bit of a code smell to me. :-) Perhaps @mikekaganski has an idea.

Other than that, I don't see anything obviously problematic with this, let's get this in as-is, given that it's already in master.

Reply:
Agreed in general. The challenge here is that the arguments are enum entries. They aren't variables and they don't exist yet, so we can't reference them. The macros are used to duplicate the entries into the original `enum { ... }` clause and a `const char* [] = { ... }` declaration.

Reflections (in C++23?) might make this easier, but I don't know of any other way. Happy to hear suggestions.